picasso/src/main/java/com/squareup/picasso3/ImageDecoderFactory.java

```diff
@@ -0,0 +1,28 @@
+package com.squareup.picasso3;
+
+import android.support.annotation.Nullable;
+import java.util.List;
+import okio.BufferedSource;
+
+final class ImageDecoderFactory {
+
+  private final List<ImageDecoder> decoders;
+
+  ImageDecoderFactory(List<ImageDecoder> decoders) {
+    this.decoders = decoders;
+  }
+
+  /**
+   * Returns the first {@link ImageDecoder} that can handle the supplied <code>source</code>.
+   * @param source The source of the image data.
+   * @return The first ImageDecoder that can decode the source, or null.
+   */
+  @Nullable ImageDecoder getImageDecoderForSource(BufferedSource source) {
+    for (ImageDecoder decoder : decoders) {
```

Comment:
I wasn't sure if we should shortcut and just return if there is only one decoder, as in a user added `.asBitmap()` to a request.  That would remove the need to check if the decoder can handle the source, but might also make the error message more confusing since it will move the error from here into the `decode` method.

Reply:
it should always call canHandleSource, at least to avoid surprise.